### PR TITLE
Remove n^2 algorithm in CodeWriter.resolve() by precomputing all of the nested simple names of a TypeSpec

### PR DESCRIPTION
**Pull request by ronshapiro**
This is the KotlinPoet equivalent of https://github.com/square/javapoet/pull/705

### CODE REVIEW

File: src/main/java/com/squareup/kotlinpoet/TypeSpec.kt [hunk 1]
```diff
@@ -277,6 +278,14 @@ class TypeSpec private constructor(builder: TypeSpec.Builder) {
     }
   }
 
+  private fun nestedTypesSimpleNames(): Set<String> {
+    var nestedTypesSimpleNames = HashSet<String>(typeSpecs.size)
+    for (typeSpec in typeSpecs) {
+      nestedTypesSimpleNames.add(typeSpec.name)
+    }
+    return nestedTypesSimpleNames
+  }
```
Comment (JakeWharton) [inline src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: This function can be replaced by `typeSpecs.map { it.name }.toSet()`

File: src/main/java/com/squareup/kotlinpoet/CodeWriter.kt [hunk 2]
```diff
@@ -456,10 +456,8 @@ internal class CodeWriter constructor(
     // Match a child of the current (potentially nested) class.
     for (i in typeSpecStack.indices.reversed()) {
       val typeSpec = typeSpecStack[i]
-      for (visibleChild in typeSpec.typeSpecs) {
-        if (visibleChild.name == simpleName) {
-          return stackClassName(i, simpleName)
-        }
+      if (typeSpec.nestedTypesSimpleNames.contains(simpleName)) {
```
[the remark this answers is not in the document]
Comment (ronshapiro) [reply src/main/java/com/squareup/kotlinpoet/CodeWriter.kt]: Thanks, I appreciate the kotlin idiom comments!

File: src/main/java/com/squareup/kotlinpoet/TypeSpec.kt (hunk 1 above)
Comment (ronshapiro) [reply src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: Done :)

**Comment by Egorand** [comment]
Thanks!

**Pull request merged**